src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -824,6 +899,18 @@ private static uint HandleEventShutdownComplete(State state, ref StreamEvent evt
                 state.ShutdownCompletionSource.SetResult();
             }
 
+            // Dispose was called before complete event.
+            bool releaseHandles = Interlocked.Exchange(ref state.ShutdownDone, 2) == 1;
+            if (releaseHandles)
+            {
+                state.ShutdownState = ShutdownState.Finished;
```

Comment:
This could be extracted into some clean up helper and shared with the part from `Dispose`.

Reply:
updated.